kqueen/engines/heat.py

```diff
@@ -0,0 +1,178 @@
+        template_file, files = self._get_template_and_files(template_name)
+        try:
+            self.connection.create_stack(
+                name=name,
+                template_file=template_file,
+                files=files,
+                ssh_key_name=ssh_key_name,
+                server_image=server_image,
+            )
+        except openstack.exceptions.BadRequestException as e:
+            # NOTE(sskripnick)
+            # handle this exception separately because useful information
+            # may be found only in e.details
+            logger.exception("Failed to create stack: %s" % e.details)
+            return False, e.details
+        except Exception as e:
+            logger.exception("Failed to create stack:")
+            return False, e        
+    
+    def get(self, stack_name):
+        return self.connection.get_stack(stack_name)
+
+    def delete(self, name):
+        return True, None
+
+
+class OpenstackHeatEngine(BaseEngine):
+    name = "openstack_heat"
+    verbose_name = "Openstack Heat Engine"
+    parameter_schema = {
+        "cluster": {
+            "template": {
+                "type": "select",
+                "label": "Template",
+                "choices": (
+                    ("default", "Heat-Kubernetes"),
+                    ("empty", "Empty"),
+                ),
+            }
+        },
+        "provisioner": {
+            "auth_url": {
+                "type": "text",
+                "label": "Auth URL",
+                "order": 10,
+                "validators": {
+                    "required": True,
+                },
+            },
+            "username": {
+                "type": "text",
+                "label": "Username",
+                "order": 20,
+                "validators": {
+                    "required": True,
+                },
+            },
+            "password": {
+                "type": "password",
+                "label": "Password",
+                "order": 30,
+                "validators": {
+                    "required": True,
+                },
+            },
+            "domain_name": {
+                "type": "text",
+                "label": "Domain name",
+                "order": 40,
+                "validators": {
+                    "required": True,
+                },
+            },
+            "project_id": {
+                "type": "text",
+                "label": "Project ID",
+                "order": 50,
+                "validators": {
+                    "required": True,
+                },
+            },
+        }
+    }
+
+    def __init__(self, cluster, *args, **kwargs):
+        super().__init__(cluster, *args, **kwargs)
+        self.kwargs = kwargs
+        self.stack_name = "Kqueen_stack_" + cluster.id
+        self.heat = Heat(
+            auth_url=kwargs["auth_url"],
+            project_id=kwargs["project_id"],
+            username=kwargs["username"],
```

Comment:
kwargs.get(..., sane_default)

Reply:
All these kwargs are required (validators: reqired: true. see parameter_schema)

Reply:
yep , i know, but in gce/aks engines code we have get-case , so maybe its better to provide common arch?

Reply:
@naumvd95 So you want to remove unnecessary get calls from gce/aks? I don't mind.